New addon api versioning approach

feerrenrut opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:
Fixes #9055

### Summary of the issue:
The approach to addon compatibility checks placed a large burden on both users and addon developers.

### Description of how this pull request fixes the issue:
The compatibility check now relies on two values within NVDA, `addonAPIVersion.CURRENT` and `addonAPIVersion.BACK_COMPAT_TO`, as well as the two values expected in the addon manifest, `lastTestedNVDAVersion` and `minimumNVDAVersion`. There needs to be an overlap in these version ranges for the addon to be considered compatible.

The concept of "untested, but potentially still works" is no longer permitted. As such, the user is no longer able to "whitelist" an addon. When an addon is considered incompatible, the addon itself must be updated and re-installed. However, due to the introduction of a second value within NVDA (`addonAPIVersion.BACK_COMPAT_TO`) which will only need to be updated when there are breaking API changes (expected to happen rarely), there is no longer the requirement for the addons to be updated for each release.

This PR also stops the error sound from being heard on developer builds of the launcher (access denied error in audio ducking). This was quite distracting while testing and is essentially an expected error.

### Testing performed:
- Installed (modified) addons.
- Checked NVDA installation with incompatible addons
- Checked NVDA update check (manual & automatic) with incompatible addons.

### Known issues with pull request:
- There is no longer any warning about incompatible addons during the startup of NVDA. This means that if NVDA is updated from a second windows account (on a shared computer), then a user may have no warning that their addons are incompatible. For now, this will not be addressed.
- The format for the API version string is now stricter (to reduce the chance of ambiguity in the API version being referred to), as such any Addons that were re-released and did not specify `year.month.minor` (with all parts present) will need to be re-released. 

### Change log entry:
The current changelog entry could be slightly modified from:
```
- Add-on authors now can enforce a minimum required NVDA version for their add-ons. Unsupported add-ons will not load or install. (#6275)
- Add-on authors must now specify the last version of NVDA the add-on has been tested against. Users will be warned against installing or enabling untested add-ons. (#6275)
```
to
```
- Add-on authors now can enforce a minimum required NVDA version for their add-ons. Unsupported add-ons will not load or install. (#6275, #9055)
- Add-on authors must now specify the last version of NVDA the add-on has been tested against. Along with the minimum required NVDA version a compatibility range is formed which must overlap with the range that NVDA supports for the addon to load or install. (#6275, #9055)
```

feerrenrut committed:
Move version string regex add tests

feerrenrut committed:
Replace compatibility checks

NVDA now uses two values to determine addon compatiblilty.

feerrenrut committed:
Keep API versions as tuples

Read from addon manifest and validate and convert to tuple.
When checking for NVDA updates, convert to tuple and save state as
tuple.

This saves confusion about whether its a string or a tuple.
Fix a few other bugs.

feerrenrut committed:
Dont show incompat addons dialog on startup

feerrenrut committed:
fix logic error: enable show incompat addons dialog button

feerrenrut committed:
Remove need to manually call destroy on IncompatibleAddonsDialog

feerrenrut committed:
Fix freeze on exit when destroy not called manually for AutoWidthColumnCheckListCtrl


Clean up of the accPropServices needs to happen when the control is
destroyed. The Destroy method is not called 
by the wx framework, but
we can register to receive the event.
https://github.com/wxWidgets/Phoenix/issues/630

feerrenrut committed:
Ensure confirmation checkbox has focus for all installer warnings

feerrenrut committed:
Set a new, more accurate label

feerrenrut committed:
Fix wording of button

feerrenrut committed:
Re-order params and make them not optional.

All usages of ConfirmAddon specify the argument with a keyword so its safe to do this.

feerrenrut committed:
Add an "addon information" button to the incompatible addons list dialog

Extracts out common showAddonInfo function.

feerrenrut committed:
Removes unnecessary code

feerrenrut committed:
Validate manifest condition: minRequiredVersion <= lastTested

It's not sensible to have a minRequiredVersion later than the version
tested against.

feerrenrut committed:
update the user guide

feerrenrut committed:
remove unnecessary check for manifest errors

This is already checked (and an AddonError raised) when creating the Addon

feerrenrut committed:
Raise error for failed audioducking unless its access denied

feerrenrut committed:
improve consitency of messages for incompatible addons

feerrenrut committed:
Further consistency of wording in GUI

leonardder commented:
> ### Known issues with pull request:
>     * There is no longer any warning about incompatible addons during the startup of NVDA. This means that if NVDA is updated from a second windows account (on a shared computer), then a user may have no warning that their addons are incompatible. For now, this will not be addressed.

Have you considered that this case also applies when one manually downloads a new version of NVDA manually and runs the particular snapshot manually? I'm afraid this will happen:

* One is at NVDA 2019.1
* NVDA 2019.2 is manually downloaded, i.e. because a proxy blocks the automatic update approach. @babbagecom has clients suffering from this.
* The 2019.2 snapshot is launched and incompatible add-ons are disabled
* The add-ons state is saved
* When the launcher is closed and 2019.1 is started, the add-ons will still be disabled, since the launcher touched the add-on state.

Would it help to disable the add-ons manager on launcher copies altogether and disallow launcher copies to touch the state? Alternatively, we can just disable the enable/disable functionality from the add-ons manager when the launcher is active?

leonardder commented on source/gui/addonGui.py:
```diff
@@ -38,11 +38,11 @@ def promptUserForRestart():
-			errorDialog=False,
-			showAddonInfoFunction=None
+			showAddonInfoFunction,
+			errorDialog=False
```
I wonder whether this is possible:

Default message boxes play a Windows sound to indicate whether there is a question or an error. In wx, this behavior is controlled with wx.ICON_WARNING and wx.ICON_ERROR, respectively. I wonder whether adding one of these to the style of the dialog will also trigger this behavior or will at least display the appropriate icon.

feerrenrut replied:
No I don't think so. I currently have code to display the right icon, but if we wish to get the windows error sound, we could do something like: 
https://stackoverflow.com/questions/13836313/how-can-i-use-the-native-error-icon-and-error-sound-in-windows-using-wxpython

leonardder replied:
That would work.

* For an error: `winsound.PlaySound("SystemHand",winsound.SND_ALIAS | winsound.SND_ASYNC)`
* For a question: `winsound.PlaySound("SystemQuestion",winsound.SND_ALIAS | winsound.SND_ASYNC)`

leonardder commented on source/gui/addonGui.py:
```diff
@@ -633,114 +618,123 @@ def __new__(cls, *args, **kwargs):
+			"The following add-ons are incompatible with NVDA version {}."
+			" These add-ons can not be enabled."
+			" Please contact the add-on author for further assistance."
+		).format("%s.%s.%s" % APIVersion)
```
Same feedback as earlier, this shows version strings like 2019.2.0

leonardder commented on source/gui/addonGui.py:
```diff
@@ -633,114 +618,123 @@ def __new__(cls, *args, **kwargs):
+		):
+			# Translators: A message to explain the reason an addon is not compatible. The place holders will be replaced
+			# with Year.Month.Minor. EG 2019.1.0
+			return "An updated version of NVDA is required. NVDA version {}.{}.{} or later".format(
```
This is not translatable. should be wrapped in a _(...)

leonardder commented on source/gui/addonGui.py:
```diff
@@ -633,114 +618,123 @@ def __new__(cls, *args, **kwargs):
+		):
+			# Translators: A message to explain the reason an addon is not compatible. The place holders will be replaced
+			# with Year.Month.Minor. EG 2019.1.0
+			return (
```
```suggestion
			return _(
```

LeonarddeR reviewed (changes requested):
Not yet fully reviewed, will have more time later today.

feerrenrut commented on source/addonHandler/__init__.py:
```diff
@@ -76,30 +76,19 @@ def getRunningAddons():
+def getIncompatibleAddons(
+		currentAPIVersion=addonAPIVersion.CURRENT,
+		backCompatToAPIVersion=addonAPIVersion.BACK_COMPAT_TO):
+	""" Returns a generator of the add-ons that are not compatible.
```
The doc string? It now comes right after the function definition, but because the argument names and default values are long they are on a new line each. I could put both the args on one line, but it would be quite long.

leonardder replied:
Ugh, never mind, that's what happens if you review code without braille. I thought the parameters were variable assignments inside the function's body.

feerrenrut commented on source/audioDucking.py:
```diff
@@ -54,13 +54,25 @@ def __del__(self):
-			_lastDuckedTime=time.time()
-		else:
-			oledll.oleacc.AccSetRunningUtilityState(ATWindow,ANRUS_ducking_AUDIO_ACTIVE|ANRUS_ducking_AUDIO_ACTIVE_NODUCK,ANRUS_ducking_AUDIO_ACTIVE_NODUCK)
+		try:
```
Not really related, it just kept tripping me up when testing locally. Hearing the error, and having to find out that it was just audio ducking. I can pull this out into a separate PR if you would like?

leonardder replied:
Ah ok. I'm perfectly fine with having this solved in this pr, if you could update the initial description saying that it fixes something. It would also help to know where this access denied error is coming from and why you're ignoring it.

feerrenrut commented on source/gui/addonGui.py:
```diff
@@ -117,13 +105,54 @@ def __init__(self, parent, title, message,
+	if not versionTuple:
+		return default
+	try:
+		year, month, minor = versionTuple
```
You are totally right, not sure how I've done that.

feerrenrut commented on source/addonHandler/__init__.py:
```diff
@@ -676,11 +661,16 @@ def __init__(self, input, translatedInput=None):
 		if result != True:
 			self._errors = result
+		elif True != self._validateApiVersionRange():
+			self._errors = "Constraint not met: minimumNVDAVersion ({}) <= lastTestedNVDAVersion ({})".format(
```
`result` can be `True` or the error string, `_errors` can be `None` or the error string.

leonardder commented on source/updateCheck.py:
```diff
@@ -640,9 +645,6 @@ def _download(self, url):
 			raise RuntimeError("Content has incorrect file hash")
 		# getFileVersionInfo will fail as long as the file is still open.
```
This is no longer in used. Yet, it makes sense to close local I think.

leonardder commented on source/updateCheck.py:
```diff
@@ -745,14 +753,15 @@ def initialize():
 			"dontRemindVersion": None,
 			"pendingUpdateVersion": None,
-			"pendingUpdateVersionTuple": None,
+			"pendingUpdateAPIVersion": None,
```
in setStateToNone, you're setting this to (0, 0, 0), whereas here, this is set to None. Would it help to make this consistent?

leonardder commented on user_docs/en/userGuide.t2t:
```diff
@@ -107,10 +107,10 @@ There are also a few options in this dialog which are explained below.
-If you have add-ons already installed there may also be a warning that untested add-ons will be disabled.
++++ Incompatible add-ons warning +++[InstallWithIncompatibleAddons]
+If you have add-ons already installed there may also be a warning that incompatible add-ons will be disabled.
 Before you're able to press the Continue button you will have to use a checkbox to confirm that you understand that these add-ons will be disabled.
```
Hmm, may be change use a checkbox into check a checkbox?

feerrenrut replied:
I think check a checkbox sounds a little strange. It probably should be "use the checkbox".

leonardder replied:
Using a check box sounds a bit neutral to me. Either checking or unchecking it could be treated as usage of it.

How about enable the check box?

feerrenrut commented on source/gui/nvdaControls.py:
```diff
@@ -121,7 +120,7 @@ def __init__(self, parent, id=wx.ID_ANY, autoSizeColumnIndex="LAST", pos=wx.Defa
 		server = ListCtrlAccPropServer(self)
 		accPropServices.SetHwndPropServer(
-			hwnd=self.Handle,
+			hwnd=self.GetHandle(),
```
It was fairly incidental while I was trying to work out how to fix the issue, I can change it back if you prefer it the other way.

leonardder replied:
While I think self.Handle is more pythonic, much code within NVDA's codebase uses wx getters directly, so I don't mind at all.

feerrenrut commented on source/gui/nvdaControls.py:
```diff
@@ -134,6 +133,9 @@ def __init__(self, parent, id=wx.ID_ANY, autoSizeColumnIndex="LAST", pos=wx.Defa
 		# Register an additional event handler to call sendCheckListBoxEvent for mouse clicks if appropriate.
 		self.Bind(wx.EVT_LEFT_DOWN, self.onLeftDown)
+		# clean up of the accPropServices needs to happen when the control is destroyed. The Destroy method is not called
+		# by the wx framework, but we can register to receive the event. https://github.com/wxWidgets/Phoenix/issues/630
```
Hmm, I agree it looks better. But it assumes that we are talking about github, then again we have #issuenumber all throughout the code already. Perhaps its a safe assumption the code will remain on github, or someone will update these comments. I'll change it.

feerrenrut commented on source/gui/addonGui.py:
```diff
@@ -117,13 +105,54 @@ def __init__(self, parent, title, message,
+		return default
+	try:
+		year, month, minor = versionTuple
+		return "{year}.{month}.{minor}".format(year=year, month=month, minor=minor)
```
what would you suggest we print for the 0.0.0 tuples?

leonardder replied:
Something like unknown, indefinite or undetermined? It should probably be translatable.

feerrenrut committed:
Review actions for #9151

feerrenrut commented:
> since the launcher touched the add-on state.

Yes, thanks for reminding me. I'll look into what we can do about this.

feerrenrut committed:
Extract version formatting into functions

Review actions for #9151

feerrenrut committed:
use `_setStateToNone(state)` to initialise state dict

More consistent initialisiation.
There is also nowhere that is checking for None
Review actions for #9151

feerrenrut committed:
Fix userguide wording.

Review actions for #9151

feerrenrut committed:
Audio ducking failed warning

Ensure that a warning is put in the log for access denied errors when
setting audio ducking.

feerrenrut committed:
Addon install warning / error dialog sounds

Use windows sounds for addon install warning / error.

feerrenrut committed:
Use context manager for writing to file.

There were several ways that the file could fail to be closed.

feerrenrut committed:
Stop incompatible addons from mutating the disabled state of addons.

Incompatible addons will can not run, but their state is never set
to "disabled"

LeonarddeR reviewed (changes requested):
Sorry for again, some comments.

feerrenrut committed:
fix translator comment errors

feerrenrut commented on source/audioDucking.py:
```diff
@@ -63,13 +63,19 @@ def _setDuckingState(switch):
 			else:
 				oledll.oleacc.AccSetRunningUtilityState(ATWindow,ANRUS_ducking_AUDIO_ACTIVE|ANRUS_ducking_AUDIO_ACTIVE_NODUCK,ANRUS_ducking_AUDIO_ACTIVE_NODUCK)
 		except WindowsError as e:
+			# When the NVDA build is not signed, audio ducking fails with access denied.
```
I did look for this, but did not find any easy way to put this check in. Figured it was easier to try, and ensure the error is an expected one.

leonardder replied:
Ok. I guess it's ok to leave it as is.

feerrenrut commented on source/gui/nvdaControls.py:
```diff
@@ -203,3 +205,109 @@ def _cleanup(self):
+		DIALOG_TYPE_WARNING: wx.ART_WARNING,
+	}
+
+	_DIALOG_TYPE_SOUND_ID_MAP = {
```
Yes, this is expected, since we will leave the dialog without a sound, and with the default icon. I can add a comment to that effect.

feerrenrut commented on source/gui/addonGui.py:
```diff
@@ -39,85 +39,73 @@ def promptUserForRestart():
+		)
+		yesButton.SetDefault()
+		yesButton.Bind(wx.EVT_BUTTON, lambda evt: self.EndModal(wx.YES))
+		self.yesButton = yesButton
```
No. I will remove this line.

feerrenrut committed:
Review actions #9151

feerrenrut commented on source/addonHandler/__init__.py:
```diff
@@ -334,19 +328,27 @@ def completeRemove(self,runUninstallTask=True):
+		_blockedAddons.discard(self.name)
 		saveState()
 
 	def addToPackagePath(self, package):
```
> A first call of this function would also yield add-ons that ought to be blocked.

I don't think this is true. Not that in `_getAvailableAddonsFromPath` the `addon.name` is added to `_blockedAddons` when the addon is incompatible.  These addons then return `False` from `isRunning()` because they are blocked. `isRunning()` is the filter function used with `getRunningAddons`

leonardder replied:
Ah yes, you're right. Still, it might be good to know what you think of my proposed alternative.

feerrenrut replied:
I'm looking into your suggestion now.

Since it may be handy to refer back to, my understanding of how addons are enabled is thus:

Various modules which need to collect and call methods on the addons or other global plugins (to make them "run") do by getting the `globalPluginHandler.runningPlugins`. This `runningPlugins` set is created in `globalPluginHandler.initialize` by iterating over `listPlugins()` which seems to be iterating on modules in `globalPlugins.__path__`. I think this is initially populated by `config.addConfigDirsToPythonPackagePath(globalPlugins)` which iterates over all the "runningAddons" and calls `addon.addToPackagePath`

This seems to be why excluding an addon (eg by name), from `addon.addToPackagePath` means it will not be run.

I don't think this PR is a good place to address this, but It would be great if there was a more direct and obvious mechanism for this. For instance a function "loadAndStartAddons". It's still not clear to me that this is all that is required for the addons to run, and perhaps there is some other special treatment.

feerrenrut replied:
Yes, I think we both agree that this could be tidied up, and I think the approach you suggested would be a good place to start. It would certainly make it much more explicit about what is happening. One concern is that there are modules other than "appModules, brailleDisplayDrivers, synthDrivers, globalPlugins" that addons rely on being added to the package path.

Perhaps "run" is poor word to use, but it's suggested by functions like `getRunningAddons` and `addon.isRunning`, perhaps "registered" is better way of referring to this action, or as you have suggested (perhaps accidentally) "activated". I almost want to call it "installed", except for the confusion that creates with the copying of files".

feerrenrut replied:
>  One concern is that there are modules other than "appModules, brailleDisplayDrivers, synthDrivers, globalPlugins" that addons rely on being added to the package path.

Reading this back, I think I should clarify what I meant. I am concerned that addons may exist that rely on other modules being added to the package paths. If we restrict this to only look for a specific set of modules, we may break addons in a new way.

I'm going to mark this conversation as resolved, and try to push forward with getting this PR merged.

feerrenrut committed:
Try to reduce confusion in the addToPackagePath method

No longer modify _blockedAddons here, however still query isBlocked.
We rely on `_getAvailableAddonsFromPath` to be setting the blocked state.

LeonarddeR reviewed (approved):
Code looks ok now. @michaeldcurran: could you have an additional look?

feerrenrut commented on source/addonHandler/__init__.py:
```diff
@@ -337,15 +337,12 @@ def completeRemove(self,runUninstallTask=True):
 	def addToPackagePath(self, package):
 		""" Adds this L{Addon} extensions to the specific package path if those exist.
-		This allows the addon to run.
+		This allows the addon to "run" / be available via `globalPluginHandler.runningPlugins`.
```
I think it's the same mechanism for braille drivers?

In `braille.initialize()` there is a call to `config.addConfigDirsToPythonPackagePath(brailleDisplayDrivers)` which, for each addon in `addonHandler.getRunningAddons()` calls `addonHandler.addToPackagePath()`. This updates the `brailleDisplayDrivers__path__`

Perhaps I should extend this comment to also mention all the modules that call `config.addConfigDirsToPythonPackagePath`: 
- `globalPlugins`
- `appModules`
- `synthDrivers`
- `brailleDisplayDrivers`


Perhaps something like:

```suggestion
		This allows the addon to "run" / be available because the package is able to search its path,
		looking for particular modules. This is used by the following:
		- `globalPlugins`
		- `appModules`
		- `synthDrivers`
		- `brailleDisplayDrivers`
```

leonardder replied:
Yes, this exactly describes what it does now.

feerrenrut committed:
Improve comment in `addToPackagePath`

LeonarddeR reviewed (changes requested):
Oops, after further investigation, it looks like I missed some things.

lukaszgo1 commented:
@feerrenrut Regarding Alpha snapshot users You wrote on the development list :
>
> Some of the work-arounds, none of which I like very much might be:
> - Set the last tested value to the alpha version. I don't like this because it's essentially saying something not true, the alpha version is a moving target. An addon author would have to testing and re-releasing their addon almost constantly, and the users would have to installing the updates too to avoid instability.
> - Do nothing, accept that there may be (very few) releases where snapshot builds are not compatible with addons.
> - Have an advanced option for disabling the compatibility check for only the currently being used pre-release NVDA version.

I believe this needs to be implemented reasoning below.

>
> I don't think we need to solve this part of the problem immediately. The majority of users (those using stable releases, wanting stable addons) are well catered for.

>

 I believe this part of the problem needs to be solved before this goes to master because the entire fact of changed approach is ad least partially due to feedback of Alpha users. If they wouldn't be able to use they add-ons they would do one of the two thing:
1. Overwrite the values in the manifest by hand not something which we want in my opinion or
2. stop using Alpha builds all together which is even worse, because NV Access need feedback from snapshot users and the need for it would increase with each API breaking change.

zstanecic commented:
And… there is another keypoint.

We should ignore synth drivers for now from this change, as some developers, like code factory, are not informed of this change.

 

From: Łukasz Golonka <notifications@github.com> 
Sent: Sunday, January 20, 2019 4:45 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [nvaccess/nvda] New addon api versioning approach (#9151)

 

@feerrenrut <https://github.com/feerrenrut>  Regarding Alpha snapshot users You wrote on the development list :

Some of the work-arounds, none of which I like very much might be:

*	Set the last tested value to the alpha version. I don't like this because it's essentially saying something not true, the alpha version is a moving target. An addon author would have to testing and re-releasing their addon almost constantly, and the users would have to installing the updates too to avoid instability.
*	Do nothing, accept that there may be (very few) releases where snapshot builds are not compatible with addons.
*	Have an advanced option for disabling the compatibility check for only the currently being used pre-release NVDA version.

I believe this needs to be implemented reasoning below.

I don't think we need to solve this part of the problem immediately. The majority of users (those using stable releases, wanting stable addons) are well catered for.

I believe this part of the problem needs to be solved before this goes to master because the entire fact of changed approach is ad least partially due to feedback of Alpha users. If they wouldn't be able to use they add-ons they would do one of the two thing:

1.	Overwrite the values in the manifest by hand not something which we want in my opinion or
2.	stop using Alpha builds all together which is even worse, because NV Access need feedback from snapshot users and the need for it would increase with each API breaking change.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9151#issuecomment-455877077> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKohk26lSgs9jV03_oY4FrpUhtP0Wv42ks5vFI8QgaJpZM4Z9Tpb> .  <https://github.com/notifications/beacon/AKohk-WkMWiDgzTjGtX0YcAq8cd47J-Mks5vFI8QgaJpZM4Z9Tpb.gif>

lukaszgo1 commented:
@zstanecic wrote:
> 
> And… there is another keypoint. We should igynth drivers for now from this change, as some developers, like code factory, are not informed of this change. 

Yes they are. It is add-on author responsibility to follow screen reader developmennt not the other way around. We cannot afford making exceptions for some add-ons and not for others.

josephsl commented:
Hi,

Agreed with Lucas. This is another reason why people should keep an eye on #7105, especially add-on authors later - once we move to Python 3, there's no going back, and add-on authors should be somewhat aware of consequences of it by now.

Thanks.

michaelDCurran commented:
If we feel this might be still used for development, then we could consider adding an extra commandline argument to enable this.

leonardder commented:
> If we feel this might be still used for development, then we could consider adding an extra commandline argument to enable this.

When scripting corporate applications for customers, I tend to throw the appModules directly in the %appdata%\nvda folder, since the creation of add-ons causes some overhead. Good to know that I misbehaved in that case😜.

Having said that, I prefer a configuration setting over a command line parameter, but would at least prefer either of them.

> In short, calls to config.addConfigDirsToPythonPackagePath should be removed from appModuleHandler, braille, globalPluginHandler, and synthDriverHandler.

Note that the addon subsystem relies on these calls, see the bottom of `config.addConfigDirsToPythonPackagePath`, where the paths are also added for every add-on.

lukaszgo1 commented:
@michaelDCurran wrote:
> 
> 
> If we feel this might be still used for development, then we could consider adding an extra commandline argument to enable this.



It is useful not only for development, but also for more advanced users, who wants to change default NVDA behavior. I've for example created global plugin which fixes #8600 until that would be addressed in the core. Disabling tis possibility would be very bad for users.

michaelDCurran commented:
While developing I understand. But for advanced users you *should* 
package it as a versioned add-on. If this is currently too complex, then 
we should look at how to make it easier to quickly create an add-on that 
has appropriate version info.

There are people who still have unpackaged code floating around in those 
directories, and no amount of version checks we implement on add-ons are 
going to disable that.


Code distributed to clients' machines really should be in an add-on. If 
that is not performant, then we really must look at why that is.

feerrenrut commented:
Replying to @lukaszgo1
> Regarding Alpha snapshot users you wrote on the development list :

For anyone else interested, this quote comes from the NVDA devel mailing list thread, [specifically this message](https://sourceforge.net/p/nvda/lists/message/36492745/) (I recommend downloading the message as HTML by the way, since there are many block quotes which are not displayed very well in sourceforge)

> I believe this part of the problem needs to be solved before this goes to master because the entire fact of changed approach is ad least partially due to feedback of Alpha users.

No, I don't think the problem is as you are stating. The addons will only become incompatible when the internal NVDA API version is moved on.
- This won't happen in the first release including this change (2019.1).
- When it does happen, it means that the add-on is well and truly incompatible. For instance after NVDA is moved to Python 3.
- The kind of changes that require us to update the version are very rare, so it's unlikely that this will happen often.

A solution could be found on the add-on distribution side allowing NVDA alpha users to be able to install an "alpha" version of the add-on. This add-on would be targeting the unreleased version of NVDA. This essentially becomes a distribution problem. This option is essentially:

> Set the last tested value to the alpha version. I don't like this because it's essentially saying something not true, the alpha version is a moving target. An addon author would have to testing and re-releasing their addon almost constantly, and the users would have to installing the updates too to avoid instability.

Despite my dislike for this in the quoted email thread, I think it would be ok if this were very clearly alpha software, and not mixed with the stable add-ons that most would be installing.

Another possible solution might be to allow NVDA (only) users on the alpha update channel to "allow incompatible add-ons" from an advanced section of the NVDA preferences. We will have to think carefully about the implications and behavior of this.

Because there is no imminent cause for add-ons to become incompatible. I won't be addressing this concern in this PR. There is at least one full release cycle, likely more before there will be a change to the addon API version.

feerrenrut commented:
Replying to @michaelDCurran

> In short, calls to config.addConfigDirsToPythonPackagePath should be removed from appModuleHandler, braille, globalPluginHandler, and synthDriverHandler.

I am totally on board with this. However I would prefer to do this in a follow up PR.

feerrenrut commented on source/addonHandler/__init__.py:
```diff
@@ -649,23 +627,38 @@ class AddonManifest(ConfigObj):
 
+# NOTE: apiVersion:
+# Eg: 2019.1.0 or 0.0.0
+# Must have 3 integers separated by dots.
```
Dev guide, I can fix that!

> unnecessary extra complexity?

Possibly, but I think it's a trade for removing ambiguity. To make sure I'm on the same page, do you think you could explain the complexity that you are concerned about?

Consider the following options:
- Allow either 2019.1.0 or 2019.1
  - We must assume that the minor part is zero when it's missing.
  - We will probably have to detect this and fill it in during manifest validation
- Allow only 3 part (2019.1.0) as it is now
  - We have to validate the addon manifest for 3 parts
- Allow only 2 part (2019.1)
  - We still have to validate the addon manifest for 2 parts.
  - We lose the possibility of changing the addon api version in a minor release. It's unlikely that we would ever want to do this, but the possibility is still there.

feerrenrut replied:
> I could think of a case where an author marks his add-on compatible with NVDA API Version 2018.4, which would effectively make his add-on fail to load because 2018.4 translates to unknown, since it's missing the minor part.

In this case they will get a validation error. The version strings will be in the log. I should point out that any add-ons that have been updated to work with the first version of the add-on compatibility check, likely used a 2 part version string. These will need to be updated and re-released. This, I think, has the benefit that the add-on authors will reconsider how the compatibility system works and hopefully reduce errors based on assumptions made against the last implementation.

leonardder commented on source/buildVersion.py:
```diff
@@ -34,7 +34,12 @@ def _updateVersionFromVCS():
+def formatVersionString(withBuildPart = False):
+	if withBuildPart:
+		return "{y}.{M}.{m}.{b}".format(y=version_year, M=version_major, m=version_minor, b=version_build)
+	return "{y}.{M}.{m}".format(y=version_year, M=version_major, m=version_minor)
```
> I think it's a good idea to keep the concepts of api version and nvda separate.

This makes sense, as long as we explicitly state "NVDA api version" (not NVDA version) when mentioning the version number in the incompatible dialogs.

feerrenrut replied:
I've been thinking about this over the last few days, I'm questioning this a little. `addonAPIVersion.CURRENT` will always equal the version number, `addonAPIVersion.BACK_COMPAT_TO` will always be equal to a some past `addonAPIVersion.CURRENT` version. For end users it's probably easiest for these to be referred to as NVDA versions, otherwise we risk confusion along the lines of "well what version of NVDA do I have to install to get THAT API version", but most likely they will assume they are equal anyway. For add-on authors, its important to realise that its the API changes that they need to care about, rather than the NVDA versions, this is a subtle difference. I'll look at implementing changing this so that one is implemented in terms of the other. There is also the distinction of end user (E.G. in the GUI) facing version strings, and developer facing ones (exe properties, log files)

leonardder commented on source/gui/addonGui.py:
```diff
@@ -332,14 +344,9 @@ def installAddon(self, addonPath, closeAfter=False):
 			if not addonVersionCheck.hasAddonGotRequiredSupport(bundle):
```
I've noticed in Alpha builds that these new custom dialogs sometimes do not get the foreground. This can probably be fixed by calling gui.mainFrame.prePopup just before this line. We also need a call of postPopup somewhere probably, but I"m very unfamiliar with this part of the code base.

feerrenrut committed:
Remove unused function

feerrenrut committed:
Tidy addon installation

- Allow the possibility of using a parent window other than the addon
  manager. Useful in the case of installing via a shell extension,
  where gui.mainframe should be the parent.
- Remove coupling between installing an Addon and addon manager
- Reduce nesting
- Moved methods to module level. These had no dependence on instance,
  or class level members.

feerrenrut commented:
Hi @leonardder I have pushed a commit that I am fairly confident resolves the issue with installing addons from the shell integration. However, I think there are still issues with this when you are running an unsigned build. As such, please test with the following try build: https://ci.appveyor.com/project/NVAccess/nvda/builds/21975742/artifacts

In terms of unifying the way that versions are printed, I have decided to leave this as is for now. There are several different expectations for the version string formatting, trying to force these all to use the same function actually makes it unnecessarily complicated and error prone. Ideally, I would like to see the introduction of a "versionNumber" type, 4-part where build numbers matter, and 3 part elsewhere. But this is actually quite out of scope for the originally issue.

leonardder commented on source/gui/addonGui.py:
```diff
@@ -270,141 +270,10 @@ def onAddClick(self, evt):
-				# #4460: If we do this immediately, wx seems to drop the WM_QUIT sent if the user chooses to restart.
-				# This seems to have something to do with the wx.ProgressDialog.
-				# The CallLater seems to work around this.
-				wx.CallLater(1, self.Close)
```
I guess the context manager is the alternative approach for this? Are you sure that it won't introduce #4460 again?

feerrenrut replied:
Quite sure, this `self.Close` is for the `AddonManger` dialog. The call to `wx.CallLater` only happened when an instance of the `AddonManager` had to be created for the purpose of installing an addon from the shell integration. When there was an `AddonManager` dialog was already opened, or installation was triggered by the `AddonManager` then this wasn't hit. I expect part of the problem was in calling `del progressDialog` rather than `Destroy` as well. But either way, it's no longer necessary to destroy the `progressDialog` and the `AddonManger` at once.

LeonarddeR reviewed (approved):
I vote for merging this as soon as possible, fixing possible regressions it may introduce into subsequent PRs.

feerrenrut committed:
Clarified formatting of version strings.

One function is for outputting a full version string (including build part)
The other is for formatting a version string for the GUI (which will
remove the minor part when it is zero).

feerrenrut committed:
Renamed _showAddonUntestedDialog

Clarified the name of the function to report that an addon is no longer
supported by NVDA. The addon is considered "too old". As opposed to NVDA
being "too old" for a newly developed addon.

feerrenrut commented:
@leonardder I have pushed a couple of commits that hopefully simplify a few things. I would still like to see a unified approach to passing around a "version" concept. But it should not be part of this PR. 

@michaelDCurran Unless you have any objections, I will merge this tomorrow.

leonardder commented:
@feerrenrut Ugh, missed this again, but could you please update the developer guide according to the three integer versioning scheme for minimumRequired and lastTested API versions?

feerrenrut commented:
> @feerrenrut Ugh, missed this again, but could you please update the developer guide according to the three integer versioning scheme for minimumRequired and lastTested API versions?

Oops, yes I will do that now.

feerrenrut committed:
Updated developer guide.

Manifest value explanations needed updating to match the new validation requirements.

feerrenrut committed:
Update changes file for #9151
